### ISSUE TITLE
[rewrite] routing layout pattern

### ISSUE THREAD

**Issue by iamjohnlong** [lhorie/mithril.js#1286]
### Description:

I'm wrapping a layout component as the docs describe here https://github.com/lhorie/mithril.js/blob/rewrite/docs/route.md#wrapping-a-layout-component

Adding another route and component updates only once after navigating. 

### Steps to Reproduce:

https://jsfiddle.net/L1uoju3w/

### Expected:

Navigating to routes should render correct component each time.

### Actual:

First route change renders component, subsequent routing doesn't render the correct component.

**Comment by pygy** [comment]
Yet another bug that's fixed by this branch: https://github.com/pygy/mithril.js/tree/mount-based-router

A built `mithril.js` derived from it can be found [here](https://cdn.rawgit.com/pygy/mithril.js/861b71d024100e9b4b9f23fdce023b92c2d0236b/mithril.js)

Your example fixed: https://jsfiddle.net/ph1vdn4f/

@lhorie: I was on holidays for the last few days, router PRs will start to flow this evening, European time...

**Comment by iamjohnlong** [comment]
@pygy that fiddle isn't working. I linked your build in this one and it's working https://jsfiddle.net/64p7rsja/

Thanks 👍

**Comment by pygy** [comment]
Indeed, I hadn't saved the fiddle after forking / changing the build...

**Comment by iamjohnlong** [comment]
have your created a PR for this @pygy ?

**Comment by pygy** [comment]
Not yet. There are three to four independent PRs lurking in there and @lhorie would prefer if they were submitted individually... but it proves tedious to untangle them because:

- One change is a complete rewrite of `api/router.js` ( #1268 #1275)
- another one (that adresses #1277 / `onmatch`) affects the syntax of many tests
- a third one changes the timing (#1269, #1275), once again potentially affecting the tests (they could be tightened, I think)
- the last one depends on the first two (#1276).

So I'll either end up with PRs that build on one another, or guaranteed conflicts the minute one of them is merged.

**Comment by lhorie** [comment]
@pygy if it's going to be too hard to submit them separately, you can just submit them all together

**Comment by pygy** [comment]
@iamjohnlong See #1290 and #1291

**Comment by pygy** [comment]
@iamjohnlong [this is fixed](https://jsfiddle.net/885uh07o/1/) now that #1290 has been merged.

GitHub didn't auto-close this because that only works when PRs are merged into the main branch (`next` for Mithril), but feel free to close it now.

**Comment by iamjohnlong** [comment]
@pygy thanks for the hard work! I was playing around with it last night. 👍
